ext/openssl/ossl_pkey.c

```diff
@@ -446,6 +446,183 @@ pkey_generate(int argc, VALUE *argv, VALUE self, int genparam)
     return ossl_pkey_new(gen_arg.pkey);
 }
 
+#if OSSL_OPENSSL_PREREQ(3, 0, 0)
+#include <openssl/param_build.h>
+#include <openssl/core_names.h>
+
+struct pkey_from_data_alias {
+    char alias[10];
+    char param_name[20];
+};
+
+static const struct pkey_from_data_alias rsa_aliases[] = {
+    { "p",    OSSL_PKEY_PARAM_RSA_FACTOR1 },
+    { "q",    OSSL_PKEY_PARAM_RSA_FACTOR2 },
+    { "dmp1", OSSL_PKEY_PARAM_RSA_EXPONENT1 },
+    { "dmq1", OSSL_PKEY_PARAM_RSA_EXPONENT2 },
+    { "iqmp", OSSL_PKEY_PARAM_RSA_COEFFICIENT1 },
+    { "", "" }
+};
+
+static const struct pkey_from_data_alias fcc_aliases[] = {
+    { "pub_key",  OSSL_PKEY_PARAM_PUB_KEY },
+    { "priv_key", OSSL_PKEY_PARAM_PRIV_KEY },
+    { "", "" }
+};
+
+struct pkey_from_data_arg {
+    VALUE options;
+    OSSL_PARAM_BLD *param_bld;
+    const OSSL_PARAM *settable_params;
+    const struct pkey_from_data_alias *aliases;
+};
+
+static int
+add_data_to_builder(VALUE key, VALUE value, VALUE arg) {
+    if(NIL_P(value))
+        return ST_CONTINUE;
+
+    if (SYMBOL_P(key))
+        key = rb_sym2str(key);
+
+    const char *key_ptr = StringValueCStr(key);
+    const struct pkey_from_data_arg *params = (const struct pkey_from_data_arg *) arg;
+
+    for(int i = 0; strlen(params->aliases[i].alias) > 0; i++) {
+        if(strcmp(params->aliases[i].alias, key_ptr) == 0) {
+            key_ptr = params->aliases[i].param_name;
+            break;
+        }
+    }
+
+    for (const OSSL_PARAM *settable_params = params->settable_params; settable_params->key != NULL; settable_params++) {
+        if(strcmp(settable_params->key, key_ptr) == 0) {
+            switch (settable_params->data_type) {
+            case OSSL_PARAM_INTEGER:
+            case OSSL_PARAM_UNSIGNED_INTEGER:
+                if(!OSSL_PARAM_BLD_push_BN(params->param_bld, key_ptr, GetBNPtr(value))) {
+                    ossl_raise(ePKeyError, "OSSL_PARAM_BLD_push_BN");
+                }
+                break;
+            case OSSL_PARAM_UTF8_STRING:
+                StringValue(value);
+                if(!OSSL_PARAM_BLD_push_utf8_string(params->param_bld, key_ptr, RSTRING_PTR(value), RSTRING_LENINT(value))) {
+                    ossl_raise(ePKeyError, "OSSL_PARAM_BLD_push_utf8_string");
+                }
+                break;
+
+            case OSSL_PARAM_OCTET_STRING:
+                StringValue(value);
+                if(!OSSL_PARAM_BLD_push_octet_string(params->param_bld, key_ptr, RSTRING_PTR(value), RSTRING_LENINT(value))) {
+                    ossl_raise(ePKeyError, "OSSL_PARAM_BLD_push_octet_string");
+                }
+                break;
+            case OSSL_PARAM_UTF8_PTR:
+            case OSSL_PARAM_OCTET_PTR:
+                ossl_raise(ePKeyError, "Unsupported parameter \"%s\", handling of OSSL_PARAM_UTF8_PTR and OSSL_PARAM_OCTET_PTR not implemented", key_ptr);
+                break;
+            }
```

Comment:
Please handle `default` too. I believe it would not reachable today, but just in case a future OpenSSL version adds new data types.